src/Ocelot/RateLimiting/RateLimiting.cs

```diff
@@ -0,0 +1,195 @@
+
+    public RateLimiting(IRateLimitStorage storage)
+    {
+        _storage = storage;
+    }
+
+    public virtual RateLimitCounter ProcessRequest(ClientRequestIdentity identity, RateLimitOptions options)
+    {
+        RateLimitCounter counter;
+        var rule = options.RateLimitRule;
+        var counterId = GetStorageKey(identity, options);
+
+        // Serial reads and writes
+        lock (ProcessLocker)
+        {
+            var entry = _storage.Get(counterId);
+            counter = Count(entry, rule);
+        }
+
+        var expiration = ToTimespan(rule.Period); // default expiration is set for the Period value
+        if (counter.TotalRequests > rule.Limit)
+        {
+            var retryAfter = RetryAfter(counter, rule); // the calculation depends on the counter returned from CountRequests
+            if (retryAfter > 0)
+            {
+                // Rate Limit exceeded, ban period is active
+                expiration = TimeSpan.FromSeconds(rule.PeriodTimespan); // current state should expire in the storage after ban period
+            }
+            else
+            {
+                // Ban period elapsed, start counting
+                _storage.Remove(counterId); // the store can delete the element on its own using an expiration mechanism, but let's force the element to be deleted
+                counter = new RateLimitCounter(DateTime.UtcNow, null, 1);
+            }
+        }
+
+        _storage.Set(counterId, counter, expiration);
+        return counter;
+    }
+
+    /// <summary>
+    /// Counts requests based on the current counter state and taking into account the limiting rule.
+    /// </summary>
+    /// <param name="entry">Old counter with starting moment inside.</param>
+    /// <param name="rule">The limiting rule.</param>
+    /// <returns>A <see cref="RateLimitCounter"/> value.</returns>
+    public virtual RateLimitCounter Count(RateLimitCounter? entry, RateLimitRule rule)
+    {
+        var now = DateTime.UtcNow;
+        if (!entry.HasValue) // no entry, start counting
+        {
+            return new RateLimitCounter(now, null, 1); // current request is the 1st one
+        }
+
+        var counter = entry.Value;
+        var total = counter.TotalRequests + 1; // increment request count
+        var startedAt = counter.StartedAt;
+        if (startedAt + ToTimespan(rule.Period) >= now) // counting Period is active
+        {
+            var exceededAt = total >= rule.Limit && !counter.ExceededAt.HasValue // current request number equals to the limit
+                ? now // the exceeding moment is now, the next request will fail but the current one doesn't
+                : counter.ExceededAt;
+            return new RateLimitCounter(startedAt, exceededAt, total); // deep copy
+        }
+
+        var wasExceededAt = counter.ExceededAt;
+        if (wasExceededAt + TimeSpan.FromSeconds(rule.PeriodTimespan) >= now) // ban PeriodTimespan is active
```

Comment:
Might be converted to return statement...

Reply:
Are you asking to condense the code from this line into a **single** `return` statement?

Reply:
Fixed in commit https://github.com/ThreeMammals/Ocelot/pull/1592/commits/f70fc80adf5aca1c74525c836b3cb683e1051c27 ✔️